src/common/io-config/src/s3_provider.rs

```diff
@@ -0,0 +1,86 @@
+use std::hash::{Hash, Hasher};
+
+use pyo3::{
+    types::{PyBytes, PyModule},
+    PyAny, PyObject, PyResult, Python, ToPyObject,
+};
+use serde::{de::Visitor, Serialize};
+
+#[derive(Clone, Debug)]
+pub struct S3CredentialsProvider {
+    pub provider: PyObject,
+    pub hash: isize,
+}
+
+impl S3CredentialsProvider {
+    pub fn new(py: Python, provider: &PyAny) -> PyResult<Self> {
+        Ok(S3CredentialsProvider {
+            provider: provider.to_object(py),
+            hash: provider.hash()?,
+        })
+    }
+}
+
+impl PartialEq for S3CredentialsProvider {
+    fn eq(&self, other: &Self) -> bool {
+        self.hash == other.hash
+    }
+}
+
+impl Eq for S3CredentialsProvider {}
+
+impl Hash for S3CredentialsProvider {
+    fn hash<H: Hasher>(&self, state: &mut H) {
+        self.hash.hash(state);
+    }
+}
+
+impl Serialize for S3CredentialsProvider {
```

Comment:
Lets look into using this here: https://github.com/Eventual-Inc/Daft/blob/ad6640ea5f36b19db73f5935a81e009c438382e2/src/daft-scan/src/file_format.rs#L312

We may need to separate out that function to a src/common crate